src/compiler/manifest.rs

```diff
@@ -0,0 +1,633 @@
+                // written to disk.
+                *updated = true;
+            }
+        }
+        true
+    }
+}
+
+/// Environment variables that are factored into the manifest cached key.
+static CACHED_ENV_VARS: Lazy<HashSet<&'static OsStr>> = Lazy::new(|| {
+    [
+        // SCCACHE_C_CUSTOM_CACHE_BUSTER has no particular meaning behind it,
+        // serving as a way for the user to factor custom data into the hash.
+        // One can set it to different values for different invocations
+        // to prevent cache reuse between them.
+        "SCCACHE_C_CUSTOM_CACHE_BUSTER",
+        "CPATH",
+        "C_INCLUDE_PATH",
+        "CPLUS_INCLUDE_PATH",
+        "OBJC_INCLUDE_PATH",
+        "OBJCPLUS_INCLUDE_PATH",
+    ]
+    .iter()
+    .map(OsStr::new)
+    .collect()
+});
+
+/// Compute the hash key of compiler preprocessing `input` with `args`.
+#[allow(clippy::too_many_arguments)]
+pub fn manifest_hash_key(
+    compiler_digest: &str,
+    language: Language,
+    arguments: &[OsString],
+    extra_hashes: &[String],
+    env_vars: &[(OsString, OsString)],
+    input_file: &Path,
+    plusplus: bool,
+    config: DirectModeConfig,
+) -> anyhow::Result<Option<String>> {
+    // If you change any of the inputs to the hash, you should change `FORMAT_VERSION`.
+    let mut m = Digest::new();
+    m.update(compiler_digest.as_bytes());
+    // clang and clang++ have different behavior despite being byte-for-byte identical binaries, so
+    // we have to incorporate that into the hash as well.
+    m.update(&[plusplus as u8]);
+    m.update(&[FORMAT_VERSION]);
+    m.update(language.as_str().as_bytes());
+    for arg in arguments {
+        arg.hash(&mut HashToDigest { digest: &mut m });
+    }
+    for hash in extra_hashes {
+        m.update(hash.as_bytes());
+    }
+
+    for &(ref var, ref val) in env_vars.iter() {
+        if CACHED_ENV_VARS.contains(var.as_os_str()) {
+            var.hash(&mut HashToDigest { digest: &mut m });
+            m.update(&b"="[..]);
+            val.hash(&mut HashToDigest { digest: &mut m });
+        }
+    }
+
+    // Hash the input file otherwise:
+    // - a/r.h exists.
+    // - a/x.c has #include "r.h".
+    // - b/x.c is identical to a/x.c.
+    // - Compiling a/x.c records a/r.h in the manifest.
+    // - Compiling b/x.c results in a false cache hit since a/x.c and b/x.c
+    // share manifests and a/r.h exists.
+    let mut buf = vec![];
+    encode_path(&mut buf, input_file)?;
+    m.update(&buf);
+    let reader = std::fs::File::open(input_file).context("while hashing the input file")?;
+
+    let (digest, finder) = if config.ignore_time_macros {
+        (Digest::reader_sync(reader)?, TimeMacroFinder::new())
+    } else {
+        Digest::reader_sync_time_macros(reader)?
+    };
+    if finder.found_time {
+        // Disable direct mode
+        debug!("Found __TIME__ in {}", input_file.display());
+        return Ok(None);
+    }
```

Comment:
Why not do this in the else rathen than crate a dummy TimeMacroFinder? (the same question applies to earlier in the file)

Reply:
In this case, I have no idea, doing it in the `else` it just better. In the previous case, a tiny bit of logic (adding the hash) is shared, but it'll just duplicate it, it probably makes the code more readable.